hudi-client/src/main/java/org/apache/hudi/index/bloom/HoodieGlobalBloomIndex.java

```diff
@@ -98,26 +100,34 @@ public HoodieGlobalBloomIndex(HoodieWriteConfig config) {
       String recordKey = partitionRecordKeyPair._2();
       String partitionPath = partitionRecordKeyPair._1();
 
-      return indexFileFilter.getMatchingFiles(partitionPath, recordKey).stream()
-          .map(file -> new Tuple2<>(file, new HoodieKey(recordKey, indexToPartitionMap.get(file))))
+      return indexFileFilter.getMatchingFilesAndPartition(partitionPath, recordKey).stream()
+          .map(partitionFileIdPair -> new Tuple2<>(partitionFileIdPair.getRight(),
+              new HoodieKey(recordKey, partitionFileIdPair.getLeft())))
           .collect(Collectors.toList());
     }).flatMap(List::iterator);
   }
 
-
   /**
    * Tagging for global index should only consider the record key.
    */
   @Override
   protected JavaRDD<HoodieRecord<T>> tagLocationBacktoRecords(
       JavaPairRDD<HoodieKey, HoodieRecordLocation> keyFilenamePairRDD, JavaRDD<HoodieRecord<T>> recordRDD) {
-    JavaPairRDD<String, HoodieRecord<T>> rowKeyRecordPairRDD =
-        recordRDD.mapToPair(record -> new Tuple2<>(record.getRecordKey(), record));
 
-    // Here as the recordRDD might have more data than rowKeyRDD (some rowKeys' fileId is null),
-    // so we do left outer join.
-    return rowKeyRecordPairRDD.leftOuterJoin(keyFilenamePairRDD.mapToPair(p -> new Tuple2<>(p._1.getRecordKey(), p._2)))
-        .values().map(value -> getTaggedRecord(value._1, Option.ofNullable(value._2.orNull())));
+    Map<String, Pair<HoodieKey, HoodieRecordLocation>> keyMap = keyFilenamePairRDD
```

Comment:
this will bring all keys into memory and OOM the driver.. we need to rethink this.

Reply:
What about something like this instead? This will avoid collecting and maintains the same operations to what already exists.

[taglocation.txt](https://github.com/apache/incubator-hudi/files/3957588/taglocation.txt)

Reply:
Thanks @bschell. That works. Will wait for @vinothchandar to comment.